### PR TITLE
find a way to reduce amount of boilerplate code in tests

### PR DESCRIPTION
**Pull request by BLoHny**
relates https://github.com/mfvanek/pg-index-health/issues/375

### CODE REVIEW

**Comment by mfvanek** [comment]
@BLoHny 
Clone https://github.com/mfvanek/pg-index-health
Then create a branch.
Other maintainers (primarily me) will be able to add commits to this branch.

File: pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsCheckOnClusterTest.java [hunk 1]
```diff
@@ -50,12 +50,12 @@ void onDatabaseWithThem(final String schemaName) {
             assertThat(check.check(ctx, FilterTablesByNamePredicate.of(ctx.enrichWithSchema("accounts"))))
                 .isEmpty();
 
-            ExecuteUtils.executeOnDatabase(getDataSource(), statement -> {
+            ExecuteUtils.executeOnDatabase(getDataSource(), (statement, sName) -> {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/checks/cluster/NotValidConstraintsCheckOnClusterTest.java:53]: It doesn't look good. We need to leave DbStatement untouched

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java [hunk 2]
```diff
@@ -43,4 +42,18 @@ protected void throwExceptionIfTableDoesNotExist(@Nonnull final Statement statem
             throw new IllegalStateException(String.format("Table with name '%s' in schema '%s' wasn't created", tableName, schemaName));
         }
     }
+
+    protected abstract List<String> getSqlToExecute(@Nonnull String schemaName);
+
+    @Override
+    public void execute(@Nonnull final Statement statement, @Nonnull final String schemaName) throws SQLException {
+        postExecute(statement, schemaName);
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:50]: That should be after for loop

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java [hunk 3]
```diff
@@ -43,4 +42,18 @@ protected void throwExceptionIfTableDoesNotExist(@Nonnull final Statement statem
...
+        for (final String sql : getSqlToExecute(schemaName)) {
+            statement.execute(sql);
+        }
+    }
+
+    protected void postExecute(@Nonnull final Statement statement, @Nonnull final String schemaName) throws SQLException {
+        Logger.getLogger(schemaName);
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:57]: Remove this line. Just try to add a comment

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java [hunk 4]
```diff
@@ -13,18 +13,17 @@
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
-import java.util.Objects;
+import java.util.List;
+import java.util.logging.Logger;
 import javax.annotation.Nonnull;
 
 abstract class AbstractDbStatement implements DbStatement {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:20]: We need something like
```
public abstract class AbstractDbStatement implements DbStatement {
    
    private static final AtomicReference<String> SCHEMA_NAME_HOLDER = new AtomicReference<>(DEFAULT_SCHEMA_NAME);

    @Nonnull
    protected static String getSchemaName() {
        return SCHEMA_NAME_HOLDER.get();
    }
    
    public static String setSchemaName(@Nonnull final String schemaName) {
        return SCHEMA_NAME_HOLDER.getAndSet(schemaName);
    }

    // protected final String schemaName; // for removing
```

and then
```
// DatabasePopulator.java

    public void populate() {
        final String oldSchemaName = AbstractDbStatement.setSchemaName(schemaName);
        try {
            ExecuteUtils.executeInTransaction(dataSource, statementsToExecuteInSameTransaction.values());
        } finally {
            AbstractDbStatement.setSchemaName(oldSchemaName);
        }
        actionsToExecuteOutsideTransaction.forEach((k, v) -> v.run());
    }
```

**Review by mfvanek — CHANGES_REQUESTED** [review]

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateNotSuitableIndexForForeignKeyStatement.java [hunk 5]
```diff
@@ -10,19 +10,14 @@
...
 public class CreateNotSuitableIndexForForeignKeyStatement extends AbstractDbStatement {
 
-    public CreateNotSuitableIndexForForeignKeyStatement(@Nonnull final String schemaName) {
-        super(schemaName);
-    }
-
     @Override
-    public void execute(@Nonnull final Statement statement) throws SQLException {
-        statement.execute(String.format("create index if not exists " +
+    protected List<String> getSqlToExecute(@Nonnull final String schemaName) {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateNotSuitableIndexForForeignKeyStatement.java:19]: Now we can to go further and get rid of `String.format` using kind of templating `{schemaName}`

```
    protected List<String> getSqlToExecute() {
        return List.of("create index if not exists " +
            "i_accounts_account_number_client_id on {schemaName}.accounts (account_number, client_id)");
    }

```

and process the template like
```
// AbstractDbStatement.java

        final String schemaName = getSchemaName();
        for (final String sql : getSqlToExecute(schemaName)) {
            statement.execute(sql.replace("{schemaName}", schemaName));
        }
```

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateTableWithColumnOfBigSerialTypeStatement.java [hunk 6]
```diff
@@ -12,22 +12,24 @@
...
+import java.util.List;
 import javax.annotation.Nonnull;
 
 public class CreateTableWithColumnOfBigSerialTypeStatement extends AbstractDbStatement {
 
-    public CreateTableWithColumnOfBigSerialTypeStatement(@Nonnull final String schemaName) {
-        super(schemaName);
+    @Override
+    protected List<String> getSqlToExecute(@Nonnull final String schemaName) {
+        return List.of();
```
[the remark this answers is not in the document]
Comment (BLoHny) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateTableWithColumnOfBigSerialTypeStatement.java:22]: DbStatement excluding postExecute Does this mean implementation of the execute method?

[the remark this answers is not in the document]
Comment (mfvanek) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateTableWithColumnOfBigSerialTypeStatement.java:22]: I mean
```
@Override
    protected List<String> getSqlToExecute(@Nonnull final String schemaName) {
        return List.of(String.format("create table if not exists %s.bad_accounts (" +
            "id bigserial not null primary key, " +
            "name varchar(255) not null," +
            // not null constraint will be added for all serial columns
            "real_client_id bigserial," +
            "real_account_id bigserial)", schemaName));
}
```
and
```
    @Override
    public void postExecute(@Nonnull final Statement statement, @Nonnull final String schemaName) throws SQLException {
        throwExceptionIfTableDoesNotExist(statement, "bad_accounts", schemaName);
    }
```

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java (hunk 3 above)
Comment (BLoHny) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:57]: How about using @SuppressFBWarnings and @SuppressWarnings?
Errors do not disappear with simple comments.

Comment (mfvanek) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:57]: > How about using @SuppressFBWarnings and @SuppressWarnings? Errors do not disappear with simple comments.

We don't dirty the code with unnecessary annotations. We use an external config. Look at this config/spotbugs/exclude.xml

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/StatisticsAwareTestBase.java [hunk 7]
```diff
@@ -92,12 +92,12 @@ private static void sleep() {
     }
 
     protected void collectStatistics(@Nonnull final String schemaName) {
-        collectStatistics();
+        collectStatisticsWith();
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/StatisticsAwareTestBase.java:95]: Unnecessary changes. It's better to roll them back

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java [hunk 8]
```diff
@@ -13,18 +13,30 @@
...
-    protected AbstractDbStatement(@Nonnull final String schemaName) {
-        this.schemaName = Objects.requireNonNull(schemaName);
+    private static final AtomicReference<String> SCHEMA_NAME_HOLDER = new AtomicReference<>(DEFAULT_SCHEMA_NAME);
+
+    @Nonnull
+    protected static String getSchemaName() {
+        return SCHEMA_NAME_HOLDER.get();
+    }
+
+    public static String setSchemaName(@Nonnull final String schemaName) {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:31]: Let's make it `protected `as well

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java [hunk 9]
```diff
@@ -43,4 +55,19 @@ protected void throwExceptionIfTableDoesNotExist(@Nonnull final Statement statem
             throw new IllegalStateException(String.format("Table with name '%s' in schema '%s' wasn't created", tableName, schemaName));
         }
     }
+
+    protected abstract List<String> getSqlToExecute();
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:59]: Mark method with `@Nonnull`

**Comment by mfvanek** [comment]
> @BLoHny Clone https://github.com/mfvanek/pg-index-health Then create a branch. Other maintainers (primarily me) will be able to add commits to this branch.

@BLoHny 
See how to add another remote https://docs.github.com/en/get-started/getting-started-with-git/managing-remote-repositories

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java [hunk 10]
```diff
@@ -13,18 +13,30 @@
...
-import java.util.Objects;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nonnull;
 
-abstract class AbstractDbStatement implements DbStatement {
+import static io.github.mfvanek.pg.model.PgContext.DEFAULT_SCHEMA_NAME;
 
-    protected final String schemaName;
+public abstract class AbstractDbStatement implements DbStatement {
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:22]: Try to remove `public` here

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateDuplicatedIndexStatement.java [hunk 11]
```diff
@@ -10,29 +10,26 @@
...
+                "on {schemaName}.accounts (account_number, account_balance) where not deleted",
+            "create index if not exists i_clients_last_first " +
+                "on {schemaName}.clients (last_name, first_name)",
+            "create index if not exists i_clients_last_name " +
+                "on {schemaName}.clients (last_name)",
+            "create index if not exists i_accounts_id_account_number_not_deleted " +
+                "on {schemaName}.accounts (id, account_number) where not deleted"
+        );
     }
+
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateDuplicatedIndexStatement.java:34]: Unnecessary empty line. Please remove it

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateSchemaStatement.java [hunk 12]
```diff
@@ -13,16 +13,18 @@
...
+import java.util.List;
 import javax.annotation.Nonnull;
 
 public class CreateSchemaStatement extends AbstractDbStatement {
 
-    public CreateSchemaStatement(@Nonnull final String schemaName) {
-        super(schemaName);
+    @Override
+    protected List<String> getSqlToExecute() {
+        return List.of("create schema if not exists " + getSchemaName());
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateSchemaStatement.java:23]: Use placeholder `{schemaName}` here

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateSchemaStatement.java [hunk 13]
```diff
@@ -13,16 +13,18 @@
...
     }
 
     @Override
-    public void execute(@Nonnull final Statement statement) throws SQLException {
+    public void postExecute(@Nonnull final Statement statement, @Nonnull final String schemaName) throws SQLException {
         statement.execute("create schema if not exists " + schemaName);
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/CreateSchemaStatement.java:28]: Remove line `statement.execute("create schema if not exists " + schemaName);` from `postExecute`

**Review by mfvanek — CHANGES_REQUESTED** [review]
@BLoHny
`+332 −439` - nice job!!!
Need some polish

**Comment by mfvanek** [comment]
@mfaulther please take a look at this PR

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java (hunk 10 above)
Comment (BLoHny) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:22]: If setSchemaName method and AbstractDbStatement are set as protected or default as controllers, they cannot be referenced from DataBasePopulator of other packages.

Shall we introduce the facade pattern?

Comment (mfvanek) [reply pg-index-health/src/test/java/io/github/mfvanek/pg/support/statements/AbstractDbStatement.java:22]: > If setSchemaName method and AbstractDbStatement are set as protected or default as controllers, they cannot be referenced from DataBasePopulator of other packages.
> 
> Shall we introduce the facade pattern?

Oh, I missed this moment. Thank you. Let's leave it as it is

**Comment by BLoHny** [comment]
I have a question. In two specific tests, the following error occurs: Container startup failed for image docker.io/bitnami/postgresql-repmgr:16.2.0. The docker image for that tag does not exist? It's been quite long time after the error occured

**Comment by mfvanek** [comment]
> I have a question. In two specific tests, the following error occurs: Container startup failed for image docker.io/bitnami/postgresql-repmgr:16.2.0. The docker image for that tag does not exist? It's been quite long time after the error occured

Try to pull image manually https://hub.docker.com/layers/bitnami/postgresql-repmgr/16.2.0/images/sha256-49592c2cd1473011e63581090e3fbdb7aedb1a537579c024ea70f43aa76be39e?context=explore

File: pg-index-health/src/test/java/io/github/mfvanek/pg/support/StatisticsAwareTestBase.java [hunk 14]
```diff
@@ -97,7 +97,7 @@ protected void collectStatistics(@Nonnull final String schemaName) {
     }
 
     protected void collectStatistics() {
-        ExecuteUtils.executeOnDatabase(getDataSource(), statement -> statement.execute("vacuum analyze"));
+        ExecuteUtils.executeOnDatabase(getDataSource(), statement -> statement.execute("VACUUM ANALYZE"));
```
Comment (mfvanek) [inline pg-index-health/src/test/java/io/github/mfvanek/pg/support/StatisticsAwareTestBase.java:100]: Please revert this change

**Review by mfvanek — CHANGES_REQUESTED** [review]
All tests should pass. See https://github.com/mfvanek/pg-index-health/pull/411#issuecomment-2143984732

**Review by mfaulther — COMMENTED** [review]
Everything is fine except [failed job](https://github.com/mfvanek/pg-index-health/actions/runs/9348963919/job/25729239280)
Please fix it and i'll approve MR

**Comment by BLoHny** [comment]
We will close this pull request to create a pull request to a new branch.
